qp_deblur/deblur.py

```diff
@@ -157,6 +159,177 @@ def generate_sepp_placements(seqs, out_dir, threads, reference_phylogeny=None,
+    placements : dict of strings
+        keys are the seqs, values are the new placements as JSON strings
+    out_dir : str
+        The job output directory
+    reference_template : str, optional
+        Filepath to the reference placement json file.
+        This file can be produced via _generate_template_rename() and should be
+        stored in the plugin package, because it can re used.
+        If None, it falls back to the Greengenes 13.8 99% reference.
+    reference_rename : str, optional
+        Similar to reference_template, but a filepath to the generated python
+        renaming script to undo the name scaping post guppy.
+        If None, it falls back to the Greengenes 13.8 99% reference.
+
+    Returns
+    -------
+    str
+        The phylogenetic insertion tree in Newick format.
+    """
+
+    # create a valid placement.json file as input for guppy
+    file_ref_template = resource_filename(
+        Requirement.parse('qp_deblur'),
+        'qp_deblur/assets/tmpl_gg13.8-99_placement.json')
+    if reference_template is not None:
+        file_ref_template = reference_template
+    if not exists(file_ref_template):
+        raise ValueError("Reference template '%s' does not exits!" %
+                         file_ref_template)
+    with open(file_ref_template, 'r') as f:
+        plcmnts = json.loads(f.read())
+        for sequence, placement in placements.items():
+            plcmnts['placements'].append({'p': placement,
+                                          'nm': [[sequence, 1]]})
+    file_placements = '%s/placements.json' % out_dir
+    with open(file_placements, 'w') as f:
+        json.dump(plcmnts, f)
+
+    # execute guppy
+    file_tree_escaped = join(out_dir, 'insertion_tree.tre')
+    fp_guppy = _get_guppy_binary()
+    std_out, std_err, return_value = system_call(
+        '%s tog %s -o %s' % (fp_guppy, file_placements, file_tree_escaped))
+    if return_value != 0:
+        error_msg = ("Error running guppy:\nStd out: %s\nStd err: %s"
+                     % (std_out, std_err))
+        raise ValueError(error_msg)
+
+    # execute node name re-labeling (to revert the escaping of names necessary
+    # for guppy)
+    file_ref_rename = resource_filename(
+        Requirement.parse('qp_deblur'),
+        'qp_deblur/assets/tmpl_gg13.8-99_rename-json.py')
+    if reference_rename is not None:
+        file_ref_rename = reference_rename
+    if not exists(file_ref_rename):
+        raise ValueError("Reference rename script does not exits!")
+    file_tree = join(out_dir, 'insertion_tree.relabelled.tre')
+    std_out, std_err, return_value = system_call(
+        'cat %s | python %s > %s' %
+        (file_tree_escaped, file_ref_rename, file_tree))
+    if return_value != 0:
+        error_msg = (("Error running %s:\n"
+                      "Std out: %s\nStd err: %s")
+                     % (file_ref_rename, std_out, std_err))
+        raise ValueError(error_msg)
+
+    with open(file_tree, 'r') as f:
```

Comment:
I'm concern that this is sending back the full tree (in memory) vs. the path. Note that we are still deciding how this will work but normally we move fullpaths/files around.

Reply:
should be easy to change the return to just the filepath instead of the contents of the file, but I feared that deletion of the working directory might happen before the result is consumed

Reply:
Excellent point but we should be fine. Now, your comment made me realize that this command is not actually running as a Qiita command, would you mind adding and adding test (or testing via it)? [See](https://github.com/qiita-spots/qp-deblur/blob/master/qp_deblur/__init__.py#L22)

Reply:
OK. I will change to returning the filepath.
@unittests: shouldn't we wait until the archive is in place and than test tree construction for meta-analyses use-cases? @josenavas point is that we still need to define the qiita function and this function here will than be called from it.